sdk/iot/hub/src/az_iot_hub_client_methods.c

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <stdint.h>
+
+#include "az_iot_hub_client.h"
+#include <az_precondition_internal.h>
+#include <az_result.h>
+#include <az_span.h>
+
+#include <_az_cfg.h>
+
+static const az_span methods_topic_prefix = AZ_SPAN_LITERAL_FROM_STR("$iothub/methods/");
+static const az_span methods_topic_filter_suffix = AZ_SPAN_LITERAL_FROM_STR("POST/");
+static const az_span methods_response_topic_result = AZ_SPAN_LITERAL_FROM_STR("res/");
+static const az_span methods_response_topic_properties = AZ_SPAN_LITERAL_FROM_STR("/?$rid=");
+static const uint8_t hash_tag = '#';
+
+
+AZ_NODISCARD az_result az_iot_hub_client_methods_subscribe_topic_filter_get(
+    az_iot_hub_client const* client,
+    az_span mqtt_topic_filter,
+    az_span* out_mqtt_topic_filter)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_topic_filter, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_topic_filter);
+
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic_filter, methods_topic_prefix, &mqtt_topic_filter));
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic_filter, methods_topic_filter_suffix, &mqtt_topic_filter));
+  AZ_RETURN_IF_FAILED(az_span_append_uint8(mqtt_topic_filter, hash_tag, &mqtt_topic_filter));
+
+  *out_mqtt_topic_filter = mqtt_topic_filter;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_hub_client_methods_received_topic_parse(
+    az_iot_hub_client const* client,
+    az_span received_topic,
+    az_iot_hub_client_method_request* out_request)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(received_topic, 1, false);
+  AZ_PRECONDITION_NOT_NULL(out_request);
+
+  int32_t index = az_span_find(received_topic, methods_topic_prefix);
+
+  if (index == -1) 
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  received_topic = az_span_slice(received_topic, index + az_span_length(methods_topic_prefix), az_span_length(received_topic));
+
+  index = az_span_find(received_topic, methods_topic_filter_suffix);
+  
+  if (index == -1) 
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  received_topic = az_span_slice(received_topic, index + az_span_length(methods_topic_filter_suffix), az_span_length(received_topic));
+
+  index = az_span_find(received_topic, methods_response_topic_properties);
+
+  if (index == -1) 
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  out_request->name = az_span_slice(received_topic, 0, index);
```

Comment:
Nit: Recommend that you treat `received_topic` variable as effectively immutable through this function and that when you need a new slice of it, you add a new variable name.  So like above `az_span method_name = az_span_slice(...)`.  Compilers should optimize out extra vars and to my eye at least it's clearer what var is at each state.

Reply:
I understand the request for more clarity, but I learned something cool recently and I think it's worth sharing here. It's this site that is useful to compare different compilers, called godbolt.org.

I took two functions, 

int function(int a) {
    a = a * 2;
    a = a * 3;
    return a;
}

int function2(int a) {
    int b = a * 2;
    int c = b * 3;
    return c;
}

Using ARM gcc 8.3.1 none (with flag -O3 for optmization) these were the assembly sizes:
function: 4021 bytes
function2:  **4827** bytes.

So there it actually the compiler is doing the optimization but not giving the result we are looking for, which is smaller size.
I know this is not the only compiler we will use, but it makes a case.
What do you think, @jspaith ?
 
Here is the [link to this exercise](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlICsupVs1qhkAUgBMAISnTSAZ0ztkBPHUqZa6AMKpWAVwC2tEJNJb0AGTy1MAOWMAjTMRAA2UgAdUCwuto9QxNBHz81Ohs7RyMXN3MlFQjaBgJmYgIg41ME5UxVANT0giiHZ1cPRTSMrJDOKuLSmLizAEpFVANiZA4AcgB6foBqABUATy9MIbHO4iG0LCGEV0xSIZIh1lRmdCGRIcwAD2YjL3YAOnEABgBBWwIhqgNaAroIe73WofEAdllboaAvbfADMABFgQAqIaScQg/43IHAuEQ5hDaEguEIpHETAELq0PZY643X5gkm9dqsEC9ACsvVIpl6VwZqBpOjkciGClmPW%2BkhBnAZBBpLNa7QA1mZaUIadwGUyWaQ2b0GQoQFdSCLmZTSHBYEg0Kc8OwyBQIEavCaKukjAAOQWkKgmgiudUQJyihlOWzpMY0oWkI1GLQEADytFY/p1pCwRhEwHYXtjeFxBQAbph1TGjvkDK6Awz7spk6w8E5iH69FhC1riHgjIX2jR6Ew2BwePxBMJRChOTIhOX1ZB2qgvMlswBaMMgtV5V6mCCWWqmeqWJrlNz1ML%2BOgr0K%2BXe0DexCr1RL5ZJFGr6bKCC8L68lWxlU9bho34Krj/P6KbrjtDyXQ9ABsp0gqyYqjcABKACyQzAMgyBDHa5wgucnBDBAtB0JgXwQLghAbFIjpDHoxqmvygpfByMhyMKXriqQUqSDK1K9PKjKQTSaoalqjF6jAiAoBgOD4EQZqUC2jAsEmnYCOYPZiP28gPskmjaPu5jri%2BzQVDKO7JFp3iHskJ4tDKamFNUmS3iEuRJNZjS6f%2Blk2cZCg2eZ%2BmAbyfQfE8LzJJI7z0J83x/CSSIfE4oKouiMLEgCQIfMhKJDLFGJJYiQK4vixCEhI8IkmSursfSXExiqtGyDI3J%2BVRkgMTqTEsdw5wAJxdd1PU9Z47GcYqrI8YofHamKgnwBAhqoBRrjkJQlrWm4toOvUzqsK6xDup6MY%2BrQfq1sGoYRlGyZxgmSYxvgaZqJm2ZKrmyD5n0gbFuxSplhWVaibWBD1o2vRCs2dAye2XC8ApQgJn2dEDl9w4QKO44BFOM5zo5GhLppdmrh1FjaN5bi8IZAT7pw%2BOk5ELlviAvBWXQT7k/jDMpF5NMtPT7m41wLPs3%2BtPcL5wEcJIVI0hVQ3KjSMHwYhyGoehmHYbh%2BGERJjVrORVqUSRYtkSpzVipK0pgYN3GqqNmrjZS4u9E1lVKiqRu26QmbbQEdNAA%3D%3D%3D
).

Here are more results:

Compiler: [armv7-clang 9](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlICsupVs1qhkAUgBMAISnTSAZ0ztkBPHUqZa6AMKpWAVwC2tEJNJb0AGTy1MAOWMAjTMRAA2UgAdUCwuto9QxNBHz81Ohs7RyMXN3MlFQjaBgJmYgIg41ME5UxVANT0giiHZ1cPRTSMrJDOKuLSmLizAEpFVANiZA4AcgB6foBqABUATy9MIbHO4iG0LCGEV0xSIZIh1lRmdCGRIcwAD2YjL3YAOnEABgBBWwIhqgNaAroIe73WofEAdllboaAvbfADMABFgQAqIaScQg/43IHAuEQ5hDaEguEIpHETAELq0PZY643X5gkm9dqsEC9ACsvVIpl6VwZqBpOjkciGClmPW%2BkhBnAZBBpLNa7QA1mZaUIadwGUyWaQ2b0GQoQFdSCLmZTSHBYEg0Kc8OwyBQIEavCaKukjAA3H4AWmQwlEjoAnKQqCaCK51RAnKKGU5bOkxjShaQjUYtAQAPK0VjhnWkLBGETAdhB1N4XEFO2YdUpo75Ay%2BiMM%2B7KbOsPBOYhhvRYCta4h4IwV9o0ehMNgcHj8QSusScmRCOvqyDtVBeZJFx1xkFqvKvUwQSy1Uz1SxNcpuephfx0TehXxH2i72IVeqJfLJIo1fTZQS31cPkq2MpX/cNR/BLe/h%2B0R7lw7Q8l0PSgbKdIKtmKq2g6jpoi6GZDO6QwQLghAbFIgprHoxqmvygpfByMhyMKQbiqQUqSDK1K9PKjJwTSaoalqVF6jAiAoBgOD4EQZqUN2jAsFmA4COYw4oKO8ivskmjaCe5g7p%2BzQVDKh7JMp3hnskl4tDK8mFNUmRPiEuRJCZjRqSBRmmTpCimQZGlgbyfQfE8LzJJI7z0J83x/CSSIfE4oKouiMLEgCQIfMg4VDGFGLRYiQK4vixCEhI8IkmSuoMfSzEpiqZGyDI3LucRkiUTq1G0dw5zuk1zUtS1ngMUxiqsqxijsdqYpcfAECGqghGuOQlCWtabgIU6KFup63qsL6xD%2BoGKYhrQYYttGsYJkm2ZphmWYpvgeZqAWRZKiWyBln0kZVgxSq1vWjZ8S2BBth2vRCl2dCiX2XC8JJQgZjJ5Fji9k4QNOs4BPOi7LlZGjrkp5kAapwHfqe4QBCeB56QELk/sZdDvvjigrvezm2djN4OejL401jLScG5EEcJIVI0oVXXKjSs1IfMw5oRhWGCVV%2BGjVaRG4VzQylRRHG1ZK0rQZ1LGqr1mr9ZS3O9NVRVKiqNViu0BarQEIDcEAA%3D)
function: 8285 bytes
function2: **9958** bytes

Compiler: [Arduino UNO (1.8.9)](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlICsupVs1qhkAUgBMAISnTSAZ0ztkBPHUqZa6AMKpWAVwC2tEJNJb0AGTy1MAOWMAjTMRAA2UgAdUCwuto9QxNBHz81Ohs7RyMXN3MlFQjaBgJmYgIg41ME5UxVANT0giiHZ1cPRTSMrJDOKuLSmLizAEpFVANiZA4AcgB6foBqABUATy9MIbHO4iG0LCGEV0xSIZIh1lRmdCGRIcwAD2YjL3YAOnEABgBBWwIhqgNaAroIe73WofEAdllboaAvbfADMABFgQAqIaScQg/43IHAuEQ5hDaEguEIpHETAELq0PZY643X5gkm9dqsEC9ACsvVIpl6VwZqBpOjkciGClmPW%2BkhBnAZBBpLNa7QA1mZaUIadwGUyWaQ2b0GQoQFdSCLmZTSHBYEg0Kc8OwyBQIEavCaKul0AZbJ1aKhOAAOACcpCoJoIrnVECcooZTls6TGNKFpCNRi0BAA8rRWGGdaQsEYRMB2IGU3hcQUAG6YdXJo75Aw%2B8MM%2B7KLOsPBOYihvRYCta4h4IwV9o0ehMNgcHj8QTCUQoTkyIR19WQdqoLzJIsAWljILVeVepggllqpnqlia5Tc9TC/jo29CvhPtH3sQq9US%2BWSRRq%2Bmygnv66fJVsZRvh4az%2BCHd/y/aIDy4doeS6HpwNlOkFSzFUbmIO0HSGABVJ0hggThzhdc43S%2BCBcEIDYpEFNY9GNU1%2BUFL4ORkORhUDcVSClSQZWpXp5UZBCaTVDUtWYvUYEQFAMBwfAiDNShu0YFhMwHARzGHMQx3kd9kk0bQz3MPdv2aCoZWPZIdO8C9kmvFoZQ0wpqkyF8QlyJJbMafSwOsuzTIUOzLMMiDeT6D4nheZJJHeehPm%2BP4SSRD4nFBVF0RhYkASBD5kASoZ4oxFLESBXF8WIQkJHhEkyV1Tj6R45MVXo2QZG5AKaMkJidRYtjuHwt1up63qes8TjuMVVk%2BMUATtTFYT4AgQ1UCo1xyEoS1rTcW17SdZ5nXdT1vV9SgA2TYNaFDFsoxjeNEyzVN00zZN8FzNQCyLJUS2QMs%2BgjKtOKVWt60bcSWwINsO16IUuzoOS%2By4XglKEdNRwY8dfqnCAZznAJF2XVdnI0TdtIcoC9NA39z3CAIzyPcyAl8v8bLoT8KcUNdHx8tySbvTyCbfVniZaTh/KgjhJCpGkquG5UaSQlDMIw1AsJwvCCKw4ipOaii5qtaiyOFoY6sYwS2slaVYKG3jVTGzUJspEXeha6qlRVVqxXaAtiD8DRuCAA)
function: 3408 bytes
function2: **4238** bytes

Reply:
I could try getting the numbers for the actual code in the PR, so we can have a definitive comparison.

Reply:
shout out @ahsonkhan for the website 🙏

Reply:
I would also agree with the way that Ewerton currently has this implemented. If the topic was passed as a reference, I would think otherwise. Here we can reuse it here until we either have success or fail (not for this feature). In either case, the `received_topic` will be popped off the stack and the original from the previous caller will be none the wiser.

Reply:
Agreed with Ewerton here - if there's reason to leave code as is (and there is given size) then let's leave as-is.  Thanks for the tip.  "Weird, wild stuff.  I did not know that."

Reply:
@ewertons, I am a bit confused by the results and exercise you shared.

> Using ARM gcc 8.3.1 none (with flag -O3 for optmization) these were the assembly sizes:
function: 4021 bytes
function2: 4827 bytes.

Where in the godbolt link do you see 4021 and 4827 bytes? That's **a lot** for a 2-4 line function. Where are these large numbers coming from? Also, the assembly looks identical to me in both cases.

Lastly, make sure to pick C and not C++ :)